### PR TITLE
Expose cipher digests and digest sizes

### PR DESCRIPTION
**Pull request by Ralith**

### CODE REVIEW

File: openssl/src/hash.rs [hunk 1]
```diff
@@ -16,6 +16,8 @@ use error::ErrorStack;
 pub struct MessageDigest(*const ffi::EVP_MD);
 
 impl MessageDigest {
+    pub(crate) fn from_ptr(x: *const ffi::EVP_MD) -> Self { MessageDigest(x) }
```
Comment (sfackler) [inline openssl/src/hash.rs]: We can just make this public (and unsafe).

File: openssl/src/ssl/mod.rs [hunk 2]
```diff
@@ -1814,6 +1815,13 @@ impl SslCipherRef {
             String::from_utf8(CStr::from_ptr(ptr as *const _).to_bytes().to_vec()).unwrap()
         }
     }
+
+    /// Returns the handshake digest of the cipher.
+    #[cfg(all(feature = "v111", ossl111))]
+    pub fn handshake_digest(&self) -> Option<MessageDigest> {
```
Comment (sfackler) [inline openssl/src/ssl/mod.rs]: For what ciphers will this be `None`?

File: openssl/src/ssl/mod.rs [hunk 3]
```diff
@@ -1814,6 +1815,13 @@ impl SslCipherRef {
             String::from_utf8(CStr::from_ptr(ptr as *const _).to_bytes().to_vec()).unwrap()
         }
     }
+
+    /// Returns the handshake digest of the cipher.
```
Comment (sfackler) [inline openssl/src/ssl/mod.rs]: We should note the normal things up here like the OpenSSL function and version requirements.

File: openssl/src/ssl/mod.rs (hunk 2 above)
Comment (Ralith) [reply openssl/src/ssl/mod.rs]: I have no idea, but [the function does have such a case](https://github.com/openssl/openssl/blob/master/ssl/ssl_ciph.c#L1964).

**Comment by Ralith** [comment]
Addressed comments and added a few additional trivial accessors.

File: openssl/src/ssl/mod.rs [hunk 4]
```diff
@@ -1814,6 +1815,30 @@ impl SslCipherRef {
...
+        }
+    }
+
+    /// Returns the NID corresponding to the cipher.
+    ///
+    /// Available as of OpenSSL 1.1.0. This corresponds to [`SSL_CIPHER_get_cipher_nid`]
+    ///
+    /// [`SSL_CIPHER_get_cipher_nid`]: https://www.openssl.org/docs/man1.1.0/ssl/SSL_CIPHER_get_cipher_nid.html
+    #[cfg(any(all(feature = "v110", ossl110), all(feature = "v111", ossl111)))]
+    pub fn cipher_nid(&self) -> Option<u32> {
```
Comment (sfackler) [inline openssl/src/ssl/mod.rs]: This should return an `Nid`: https://docs.rs/openssl/0.10.5/openssl/nid/struct.Nid.html

File: openssl/src/symm.rs [hunk 5]
```diff
@@ -75,6 +75,16 @@ pub enum Mode {
 pub struct Cipher(*const ffi::EVP_CIPHER);
 
 impl Cipher {
+    /// Looks up the cipher for a certain nid.
+    ///
+    /// This corresponds to [`EVP_get_cipherbynid`]
+    ///
+    /// [`EVP_get_cipherbynid`]: https://www.openssl.org/docs/man1.0.2/crypto/EVP_get_cipherbyname.html
+    pub fn from_nid(nid: u32) -> Option<Cipher> {
```
Comment (sfackler) [inline openssl/src/symm.rs]: This should take an Nid.

**Comment by sfackler** [comment]
Thanks!

**Pull request merged**